### PR TITLE
Add Expr.IsEqual and update Expr.And / Expr.Or logical operations

### PR DESCRIPTION
**Pull request by jeffmay**
- Add Expr.IsEqual for === and =!= support
  - Add simple and standard definitions
  - Add documentation
  - Update EqualComparable to use OP type
- Upgrade And / Or / Not to support wrapper types
  - Add unit tests for && / || / !
  - Add unit tests for and / or / not
  - Add logic typeclass for all logical operations
  - Update Expr.Not to match And / Or operations
  - Move unary_! definition from Expr to BuildExprDsl
  - Clean up negation documentation
  - Add documentation for Logic traits
- Add Boolean Logic tests
  - Rename and refactor ForAll and Exists tests
  - Add not() tests to SimpleLogicSpec
  - Update And / Or tests to use fixed truth tables
  - Rename StandardLogicalSpec to StandardBooleanLogicSpec
  - Add and / or tests to StandardBooleanLogicSpec
  - Split SimpleCollectionLogicSpec and cleanup tests
  - Remove redundant SimpleJustifiedCollectionLogicSpec
  - Remove redundant JustifiedCollectionEvidenceSpec
  - Add future test to SimpleJustified collection specs
  - Cleanup Simple Exists / ForAll specs
  - Add SimpleJustifiedBooleanLogicSpec
- Convert AND and OR operations to accept 2 or more expressions
  - Convert Expr.And and Expr.Or to accept a more than 2 arguments
  - Add more test cases to boolean logic specs
- Make ExtractValue extend Function1
- Add munit-scalacheck to the build

### CODE REVIEW

File: core-v1/src/main/scala/algebra/EqualComparable.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import cats.{Eq, Functor, Id, Semigroupal}
+
+trait EqualComparable[F[_], V] {
+
+  def isEqual(
+    left: F[V],
+    right: F[V],
+  ): F[Boolean]
+}
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/EqualComparable.scala]: Document and maybe add `OP[F[V]]` and `OP[F[Boolean]]` as well

File: core-v1/src/main/scala/algebra/EqualComparable.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+object EqualComparable extends LowPriorityEqualComparable {
+
+  implicit def eq[V : Eq]: EqualComparable[Id, V] = Eq[V].eqv
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/EqualComparable.scala]: Use `shapeless.Id` here instead

File: core-v1/src/main/scala/algebra/Expr.scala [hunk 3]
```diff
@@ -117,20 +117,23 @@ sealed abstract class Expr[-I, +O : OP, OP[_]](val name: String) {
...
+//    * Negate the output of this expression using the implicit definition of [[Negation]].
+//    *
+//    * @param negation the definition for how to negate the output of this expression.
+//    */
+//  def unary_![RO >: O](
+//    implicit
+//    negation: Negation[RO],
+//    opA: OP[RO],
+//  ): Expr.Not[I, RO, OP] =
+//    Expr.Not[I, RO, OP](this)
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/Expr.scala]: Remove unused code

File: core-v1/src/main/scala/logic/Conjunction.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.rallyhealth.vapors.v1
+
+package logic
+
+import shapeless.Id
+
+// TODO: Should I allow short-circuiting?
+trait Conjunction[F[_], B, OP[_]] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/logic/Conjunction.scala:24]: Document this trait

File: core-v1/src/main/scala/logic/Disjunction.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package com.rallyhealth.vapors.v1
+
+package logic
+
+import shapeless.Id
+
+// TODO: Should I allow short-circuiting?
+trait Disjunction[F[_], B, OP[_]] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/logic/Disjunction.scala:24]: Document this trait

File: core-v1/src/main/scala/logic/Logic.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package com.rallyhealth.vapors.v1
+
+package logic
+
+import shapeless.Id
+
+trait Logic[F[_], B, OP[_]] extends Conjunction[F, B, OP] with Disjunction[F, B, OP] with Negation[F, B, OP]
```
Comment (jeffmay) [inline core-v1/src/main/scala/logic/Logic.scala:22]: Document this trait

File: core-v1/src/test/scala/JustifiedSelectSpec.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    val expr = root.get(_.select(_.value))
+    val expected = Justified.byInference(
+      "select", // TODO: Put lens description here?
+      root.value.value.value,
+      NonEmptyList.of(root.value),
+    )
+    val obtained = expr.run()
+    assertEquals(obtained, expected)
+  }
+}
```
Comment (jeffmay) [inline core-v1/src/test/scala/JustifiedSelectSpec.scala]: I should add more tests here

File: core-v1/src/test/scala/example/FactTypes.scala [hunk 8]
```diff
@@ -4,11 +4,38 @@ package example
...
+
+final case class CombinedTags(
+  tags: SortedSet[String],
+  timestampMillis: Long,
+)
+
+object CombinedTags {
+
+  // TODO: Take a clock for testing?
+  def now(tags: SortedSet[String]): CombinedTags = CombinedTags(tags, Instant.now().toEpochMilli)
```
Comment (jeffmay) [inline core-v1/src/test/scala/example/FactTypes.scala]: This is an easy one to do before merging.

File: core-v1/src/main/scala/algebra/Expr.scala [hunk 9]
```diff
@@ -195,43 +215,43 @@ object Expr {
...
     *
     * @param leftExpr the left side of the `AND` operation
     * @param rightExpr the right side of the `AND` operation
     */
-  final case class And[-I, OP[_]](
-    leftExpr: Expr[I, Boolean, OP],
-    rightExpr: Expr[I, Boolean, OP],
+  final case class And[-I, +B, F[+_], OP[_]](
+    leftExpr: Expr[I, F[B], OP],
+    rightExpr: Expr[I, F[B], OP],
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/Expr.scala]: I should probably make this support 2 or more expressions. We use that in a lot of places.

**Review by vinit-garg — APPROVED** [review]

File: core-v1/src/main/scala/dsl/BuildExprDsl.scala [hunk 10]
```diff
@@ -77,6 +112,7 @@ trait BuildExprDsl extends DebugExprDsl {
     opB: OP[W[Boolean]],
   ): ComparisonExprBuilder[I, V] = new ComparisonExprBuilder(valueExpr)
 
+  // TODO: Rename to WindowComparisonExprBuilder
   class ComparisonExprBuilder[I, V : Order : OP](
     protected val valueExpr: I ~:> W[V],
   )(implicit
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/BuildExprDsl.scala]: I should do or remove this TODO

File: core-v1/src/main/scala/dsl/BuildExprDsl.scala [hunk 11]
```diff
@@ -69,6 +103,7 @@ trait BuildExprDsl extends DebugExprDsl {
     ): AndThen[I, C[W[A]], C[W[B]]]
   }
 
+  // TODO: Rename to compareWithinWindow
   implicit def compare[I, V : Order : OP](
     valueExpr: I ~:> W[V],
   )(implicit
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/BuildExprDsl.scala]: Do or remove this TODO

File: core-v1/src/main/scala/dsl/BuildExprDsl.scala [hunk 12]
```diff
@@ -116,6 +152,34 @@ trait BuildExprDsl extends DebugExprDsl {
...
+
+  class EqualComparisonExprBuilder[I, V : OP](
+    protected val leftExpr: I ~:> W[V],
+  )(implicit
+    eqV: EqualComparable[W, V, OP],
+    opV: OP[W[V]],
+    opB: OP[W[Boolean]],
+  ) {
+
+    def ===(literal: V): Expr.IsEqual[I, V, W, OP] = Expr.IsEqual(leftExpr, Expr.Const(WrapConst[W].wrapConst(literal)))
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/BuildExprDsl.scala]: Remove these literal versions to better match the `>`, `<=`, etc comparison operators

File: core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala [hunk 13]
```diff
@@ -28,13 +31,13 @@ trait JustifiedBuildExprDsl extends WrappedBuildExprDsl with JustifiedDslTypes {
...
-    negation: Negation[Justified[O]],
-  ): Expr.Not[Justified[I], Justified[O], OP] =
-    Expr.Not(expr)
+//  override final def not[I, O](
+//    expr: Justified[I] ~:> Justified[O],
+//  )(implicit
+//    opO: OP[Justified[O]],
+//    negation: Negation[Justified[O]],
+//  ): Expr.Not[Justified[I], Justified[O], OP] =
+//    Expr.Not(expr)
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala]: Remove or replace commented code if helpful

File: core-v1/src/test/scala/JustifiedSelectSpec.scala (hunk 7 above)
Comment (jeffmay) [reply core-v1/src/test/scala/JustifiedSelectSpec.scala]: It looks I need to figure out how to handle the `.get(_.select(_.someOptionalOrSeqValue))` case better. In the old DSL, you would have to use `.getFoldable`, but that shouldn't be necessary anymore.

File: core-v1/src/main/scala/debug/HasShow.scala [hunk 14]
```diff
@@ -31,6 +31,8 @@ object HasShow extends LowPriorityHasShow {
   private final case class Impl[V](show: Show[V]) extends HasShow[V]
 
   implicit def hasShow[V](implicit show: Show[V]): HasShow[V] = Impl(show)
+
+  implicit def hasShowEmpty[C[a] <: IterableOnce[a]]: HasShow[C[Nothing]] = Impl(Show.fromToString)
```
Comment (jeffmay) [inline core-v1/src/main/scala/debug/HasShow.scala]: Is this still needed? If so, document it.

File: core-v1/src/main/scala/dsl/BuildExprDsl.scala [hunk 15]
```diff
@@ -4,42 +4,86 @@ package dsl
...
-  self: DslTypes =>
+  self: DslTypes with WrapImplicits =>
+
+  protected implicit def boolLogic: Logic[W, Boolean, OP]
 
   protected implicit def windowComparable: WindowComparable[W, OP]
 
   protected implicit def extract: Extract[W]
 
+  protected implicit def wrapElement: SelectElement[W]
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/BuildExprDsl.scala]: Rename to selectElement?

File: core-v1/src/main/scala/dsl/GetAsWrapper.scala [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+import algebra.{Expr, Extract, SelectFromWrapped, WrapConst}
+import lens.VariantLens
+import cats.Functor
+import cats.implicits._
+import shapeless.Id
+
+import scala.collection.Factory
+
+// TODO: Document
+class GetAsWrapper[-I, W[+_] : Extract : WrapConst : SelectElement, A, C[_], OP[_]](inputExpr: Expr[I, W[A], OP]) {
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/GetAsWrapper.scala]: BLKR: I should definitely document this.

File: core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala [hunk 17]
```diff
@@ -2,39 +2,46 @@ package com.rallyhealth.vapors.v1
...
+import com.rallyhealth.vapors.v1.data.{FactTypeSet, Justified, Window}
+import com.rallyhealth.vapors.v1.lens.VariantLens
+import com.rallyhealth.vapors.v1.logic.Logic
+
+trait JustifiedBuildExprDsl extends BuildExprDsl with WrapJustifiedImplicits with JustifiedDslTypes {
+
+  implicit def wrapWindow[O](window: Window[O])(implicit opW: OP[Justified[Window[O]]]): Any ~:> Justified[Window[O]] =
+    Expr.Const(wrapConst.wrapConst(window))
+
+  // TODO: Remove the need for the above line ^
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala]: I should either move this to the `BuildExprDsl` or just require using `Window.between(x, y).const`... or maybe some other syntax, but still using `.const`

File: core-v1/src/main/scala/dsl/SelectElement.scala [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+package dsl
+
+import shapeless.Id
+
+import scala.annotation.implicitNotFound
+
+// TODO: Include OP type here?
+// TODO: Figure out how to make W invariant?
+trait SelectElement[W[+_]] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/SelectElement.scala]: I should document this. I don't think making `W` invariant on its parameter is necessary or really preferable, so I should remove that TODO. I played around with adding the `OP`, but I'm not sure if it is necessary. Might be worth another shot though.

File: core-v1/src/main/scala/dsl/WrapImplicits.scala [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+  implicit def id[A]: OutputType.Aux[W, A, W[A]]
+
+  /**
+    * Use this to implement all the DSL definitions
+    */
+  protected val defn: WrapDefinitions[W, OP]
+}
+
+// TODO: Remove OP type if unused
+final class WrapDefinitions[W[+_] : WrapConst : SelectElement, OP[_]] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/WrapImplicits.scala]: Do or remove TODO

File: core-v1/src/main/scala/dsl/Wrappable.scala [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+  @implicitAmbiguous(
+    """Can't use the .get operation to select a field of type ${C}[${A}].
+
+This is because there is no Functor[${C}] definition. You can use .getAs[Seq] to convert the value to a Seq before mapping over the results.""",
+  )
+  implicit def setIsNotWrappable1[C[a] <: Set[a], A]: Wrappable[C[A]] = unexpected
+  implicit def setIsNotWrappable2[C[a] <: Set[a], A]: Wrappable[C[A]] = unexpected
+
+  // TODO: Other common non-Functor output types?
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/Wrappable.scala]: Maybe I should add `Array`? It is probably rare, but common enough to warrant a special compiler error message.

Comment (jeffmay) [reply core-v1/src/main/scala/dsl/Wrappable.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/dsl/WrapImplicits.scala (hunk 19 above)
Comment (jeffmay) [reply core-v1/src/main/scala/dsl/WrapImplicits.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/dsl/SelectElement.scala (hunk 18 above)
Comment (jeffmay) [reply core-v1/src/main/scala/dsl/SelectElement.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala (hunk 17 above)
Comment (jeffmay) [reply core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/dsl/GetAsWrapper.scala (hunk 16 above)
Comment (jeffmay) [reply core-v1/src/main/scala/dsl/GetAsWrapper.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/dsl/BuildExprDsl.scala (hunk 15 above)
Comment (jeffmay) [reply core-v1/src/main/scala/dsl/BuildExprDsl.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/debug/HasShow.scala (hunk 14 above)
Comment (jeffmay) [reply core-v1/src/main/scala/debug/HasShow.scala]: Moved to https://github.com/jeffmay/vapors/pull/80

File: core-v1/src/main/scala/logic/Conjunction.scala [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+  * TODO: Should this allow short-circuiting? Maybe return an Option or Eval?
+  *
+  * @tparam F the wrapper (or effect) type over which equality is computed
+  * @tparam B the boolean-like value type to perform conjunction
+  * @tparam OP a custom output parameter type used by visitors to enable post-processing operations.
+  *            See [[dsl.DslTypes.OP]] for more details.
+  */
+@implicitNotFound(
+  """Cannot perform a logical AND operation on values of type ${F}[${B}] with an output parameter of ${OP}. 
+If you want like to define conjunction (aka the AND operation) for a custom type, you must define or import an instance of Conjunction[${F}, ${B}, ${OP}]""",
```
Comment (paul-a-kennedy-rally) [inline core-v1/src/main/scala/logic/Conjunction.scala]:
```suggestion
To define conjunction for a custom type (aka the AND operation), you must define or import an instance of Conjunction[${F}, ${B}, ${OP}]""",
```

File: core-v1/src/main/scala/logic/Disjunction.scala [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+  * TODO: Should this allow short-circuiting? Maybe return an Option or Eval?
+  *
+  * @tparam F the wrapper (or effect) type over which equality is computed
+  * @tparam B the boolean-like value type to perform conjunction
+  * @tparam OP a custom output parameter type used by visitors to enable post-processing operations.
+  *            See [[dsl.DslTypes.OP]] for more details.
+  */
+@implicitNotFound(
+  """Cannot perform a logical OR operation on values of type ${F}[${B}] with an output parameter of ${OP}. 
+If you want like to define disjunction (aka the OR operation) for a custom type, you must define or import an instance of Disjunction[${F}, ${B}, ${OP}]""",
```
Comment (paul-a-kennedy-rally) [inline core-v1/src/main/scala/logic/Disjunction.scala]:
```suggestion
To define disjunction for a custom type (aka the OR operation), you must define or import an instance of Disjunction[${F}, ${B}, ${OP}]""",
```

File: core-v1/src/main/scala/logic/Logic.scala [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  * to define without having to pass 3 type parameters to 3 separate traits.
+  *
+  * @tparam F the wrapper (or effect) type over which equality is computed
+  * @tparam B the boolean-like value type to perform the logical operations
+  * @tparam OP a custom output parameter type used by visitors to enable post-processing operations.
+  *            See [[dsl.DslTypes.OP]] for more details.
+  */
+@implicitNotFound(
+  """Cannot perform a logical operations (like AND / OR / NOT) on values of type ${F}[${B}] with an output parameter of ${OP}. 
+If you want like to define the operations, you must define or import an instance of Logic[${F}, ${B}, ${OP}]""",
```
Comment (paul-a-kennedy-rally) [inline core-v1/src/main/scala/logic/Logic.scala]:
```suggestion
To define the operations, you must define or import an instance of Logic[${F}, ${B}, ${OP}]""",
```

File: core-v1/src/main/scala/logic/Negation.scala [hunk 24]
```diff
@@ -12,40 +13,23 @@ import scala.annotation.implicitNotFound
   * @note this is different from defining how to get the negative value of a number.
   */
 @implicitNotFound(
-  "Cannot negate a value of type ${A}. If you want like to define negation (aka the NOT operation) for a custom type, you must implement or import an instance of Negation[${A}]",
+  """Cannot negate a value of type ${F}[${B}] with an output parameter of ${OP}. 
+If you want like to define negation (aka the NOT operation) for a custom type, you must define or import an instance of Negation[${F}, ${B}, ${OP}]""",
```
Comment (paul-a-kennedy-rally) [inline core-v1/src/main/scala/logic/Negation.scala]:
```suggestion
To define negation for a custom type (aka the NOT operation), you must define or import an instance of Negation[${F}, ${B}, ${OP}]""",
```

**Review by paul-a-kennedy-rally — APPROVED** [review]
LGTM

**Pull request merged**